powershell-analyzer/src/analyze.ps1

```diff
@@ -0,0 +1,215 @@
+    to "en-US".
+#>
+function Get-Locale {
+    if ($IsWindows) {
+        return (Get-WinSystemLocale).Name
+    }
+
+    $lang = (locale) -split "\n" | Where-Object { $_.startsWith("LANG=") }
+    # example: LANG=en_US.UTF-8
+
+    if ($lang.Length -eq 0) {
+        return "en-US"
+    }
+
+    return $lang.Substring(5, 5).replace('_', '-')
+}
+
+<#
+.DESCRIPTION
+    Creates the outer "envelope" for Sarif test results.
+#>
+function Get-SarifContainer {
+    return @{
+        '$schema' = "http://json.schemastore.org/sarif-2.1.0"
+        version   = "2.1.0"
+        runs      = @(
+            @{
+                tool    = @{
+                    driver = @{
+                        name     = "PSScriptAnalyzer"
+                        version  = (Find-Module PSScriptAnalyzer).Version
+                        language = Get-Locale
+                        informationUri = "https://docs.microsoft.com/en-us/powershell/module/psscriptanalyzer"
+                        rules    = @()
+                    }
+                }
+                results = @()
+            }
+        )
+    }
+}
+
+<#
+.DESCRIPTION
+    Run the PSScriptAnalyzer on a directory, adding the results into the given
+    $Sarif object.
+#>
+function Invoke-Analyzer {
+    param (
+        # Pre-created Sarif results object
+        [Parameter(Mandatory = $True)]
+        [PSCustomObject]
+        $SarifData,
+
+        # Directory to scan
+        [Parameter(Mandatory = $True)]
+        [string]
+        $Directory,
+
+        [boolean]
+        $SaveToFile
+    )
+
+    if ($null -eq $(Get-Module -ListAvailable -Name PSScriptAnalyzer)) {
+        Install-Module -Name PSScriptAnalyzer -Force
+    }
+
+    # There is a bug that causes the PSScriptAnalyzer to fail to notice the use of $SarifData inside
+    # of the ForEach-Object. And the recommended approach for suppressing messages is failing too.
+    # Therefore, this has a dummy usage of the parameter just to satisfy the analyzer
+    # https://github.com/PowerShell/PSScriptAnalyzer/issues/1472
+    $SarifData | Out-Null
+
+    if($SaveToFile) {
+        $results = Invoke-ScriptAnalyzer -Path $Directory -ExcludeRule $ExcludedRules -Recurse
+    } else {
+        Invoke-ScriptAnalyzer -Path $Directory -ExcludeRule $ExcludedRules -Recurse -ReportSummary
+        return
+    }
+
+    $results | ForEach-Object {
+        $SarifData.runs[0].results += @{
+            ruleId    = $_.RuleName
+            level     = Get-Severity $_
+            message   = @{
+                text = $_.Message
+            }
+            locations = @(
+                @{
+                    physicalLocation = @{
+                        artifactLocation = @{
+                            # Getting path of the script under analysis
+                            # https://github.com/PowerShell/PSScriptAnalyzer/issues/1758#issuecomment-1006072757
+                            uri = $_.ScriptPath
```

Comment:
This is working to get the full path of the script under analysis, but this gets the folder relative to the action, I will change it to get it relative to the repo itself so that CodeQL can show the file in the results